processor/spanmetricsprocessor/processor.go

```diff
@@ -381,37 +449,72 @@ func (p *processorImp) buildDimensionKVs(serviceName string, span pdata.Span, op
 	return dims
 }
 
-func concatDimensionValue(metricKeyBuilder *strings.Builder, value string, prefixSep bool) {
+func extractResourceAttrsByKeys(serviceName string, keys []Dimension, resourceAttrs pdata.AttributeMap) pdata.AttributeMap {
+	dims := pdata.NewAttributeMap()
+	dims.UpsertString(serviceNameKey, serviceName)
+	for _, ra := range keys {
+		if attr, ok := resourceAttrs.Get(ra.Name); ok {
+			dims.Upsert(ra.Name, attr)
+		} else if ra.Default != nil {
+			// Set the default if configured, otherwise this metric should have no value set for the resource attribute.
+			dims.Upsert(ra.Name, pdata.NewAttributeValueString(*ra.Default))
+		}
+	}
+	return dims
+}
+
+func concatDimensionValue(metricKeyBuilder *strings.Builder, value string) {
```

Comment:
I like your suggestion of simplifying this and I propose to build on this further with the following goals in mind:
- Remove the assumption that a new/empty metricKeyBuilder should be supplied to `concatDimensionValue` in order to avoid prefixing the separator.
- Remove the assumption that `metricKeyBuilder` is not `nil`.
- Remove the need to provide a reference to `metricKeyBuilder` each time, which is a means of maintaining the key building "in progress" state.
- Make the usage simpler and more intuitive with a more idiomatic function name (as well as removing the need to pass a valid `StringBuilder`).
 
So instead of:

```
	var metricKeyBuilder strings.Builder
	concatDimensionValue(&metricKeyBuilder, span.Name())
	concatDimensionValue(&metricKeyBuilder, span.Kind().String())
	concatDimensionValue(&metricKeyBuilder, span.Status().Code().String())
```

The usage could be simplified to:
```
        mkb := keybuilder.NewMetricKeyBuilder()
	mkb.Append(serviceName)
	mkb.Append(span.Name())
	mkb.Append(span.Kind().String())
	mkb.Append(span.Status().Code().String())
```

My proposal is to separate and encapsulate the concern of metrics key building from the processor via a package `processor/spanmetricsprocessor/keybuilder`:
```
package keybuilder

import "strings"

const (
	metricKeySeparator = string(byte(0))
)

type (
	MetricKeyBuilder interface {
		Append(value string)
		String() string
	}

	metricKeyBuilder struct {
		sb        strings.Builder
		separator string
	}
)

func NewMetricKeyBuilder() MetricKeyBuilder {
	return &metricKeyBuilder{}
}

func (mkb *metricKeyBuilder) Append(value string) {
	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
	mkb.sb.WriteString(mkb.separator)
	mkb.sb.WriteString(value)
	mkb.separator = metricKeySeparator
}

func (mkb *metricKeyBuilder) String() string {
	return mkb.sb.String()
}
```

Let me know what you think.

Reply:
I really like this approach, the only thing I would change is the method name of `keybuilder.NewMetricKeyBuilder()` to `keybuilder.New()` to avoid package name stuttering. 

I am fairly certain that `strings.Builder` has a `.Grow(by int)` method that we can use on the NewMethod which can define a package level default of what we consider to be a reasonable size byte slice?

Reply:
> the only thing I would change is the method name of keybuilder.NewMetricKeyBuilder() to keybuilder.New() to avoid package name stuttering.

> I am fairly certain that strings.Builder has a .Grow(by int) method that we can use on the NewMethod which can define a package level default of what we consider to be a reasonable size byte slice?

+1 to both suggestions.